Probe_Offset_Wizard: Improvements

swissnorp opened the pull request:
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description

**Config**

- Now it's possible to disable `PROBE_OFFSET_START` and use current `probe.offset.z` instead
- The set `PROBE_OFFSET_WIZARD_XY_POS` will be used for probing a `z_offset_ref` and for the nozzle/bed contact position (the nozzle will move to this position after probing).

**Procedure**

1. _Call Wizard by user_
2. Store current `probe.offset.z` for backup and overwrite it with `PROBE_OFFSET_START` if defined
3. Store Bed-Leveling-State and deactivate it
4. Home all Axes by `G28`
5. _Call Preparation_
6. Disable Soft Endstops
7. Set `z_offset_ref = 0` (for if homing is done by probe and `PROBE_OFFSET_WIZARD_XY_POS` is not   ##defined)
8. If Probe is not used for homing or `PROBE_OFFSET_WIZARD_XY_POS` is defined, `z_offset_ref` must be probed
9. If `PROBE_OFFSET_WIZARD_XY_POS` is not defined probing will be done at `XY_CENTER`
10. The Nozzle moves by the `probe.offset_xy` to the Probing point
11. _Call Menu_
12. The probed will be used to calculate the new `probe.offset.z`: `calculated_z_offset = probe.offset.z + current_position.z - z_offset_ref`
13. User adjust Offset by menu
14. _Done by user_
15. The calulated Offset is stored
16. If Probe is used for homing, homing is done with the new offset.
17. _Cancel by user_
18. The old Offset is restored
19. If Probe is used for homing and `PROBE_OFFSET_START` is set, homing is done with the restored offset.

### Benefits

- Its possible to calibrate the probe offset even if its not used for homing.
- The calibration is accurate because probing and nozzle/bed-contact points are the same
- Its possible to define a specific position for the calibration wizard

### Configurations

Niente

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/19650
https://github.com/MarlinFirmware/Marlin/issues/20204

swissnorp committed:
Update menu_probe_offset.cpp

swissnorp committed:
Update menu_advanced.cpp

swissnorp committed:
Update menu.h

swissnorp committed:
Update marlinui.h

swissnorp committed:
Update Configuration_adv.h

qwewer0 commented:
This fixed the issue 4-6 in https://github.com/MarlinFirmware/Marlin/issues/20204#issuecomment-730501360, but it still goes back to the status screen.

swissnorp committed:
Update menu_probe_offset.cpp

Add ui.defer_status_screen(); in top of goto_probe_offset_wizard()

swissnorp commented:
> 
> 
> This fixed the issue 4-6 in [#20204 (comment)](https://github.com/MarlinFirmware/Marlin/issues/20204#issuecomment-730501360), but it still goes back to the status screen.

@qwewer0 Thany you for your quick test! I could reproduce your problem. If homing takes to long (nozzle far away from endstop) it falls back to the status screen.

 I think the latest commit fixed it. Could you give it another try?

Thanks, norp

thinkyhead committed:
Change name

qwewer0 commented:
Tried it with the new changes:
And now after homing it probes the bed (`Z reference`) one more times (`3` times with `MULTIPLE_PROBING 2`, `EXTRA_PROBING 1`) while the LCD shows noting (blank), then it moves to `PROBE_OFFSET_WIZARD_XY_POS` and goes into the wizard menu and can be used as before.

- I would love to see an LCD message while it probes for the Z reference, so that it isn't just a blank screen.
- And one thing that I don't like, is the homing after the done/cancel button. Is there a particular reason why do we need to home after the wizard?

filippor commented:
Tested on a delta and it seems to work with PROBE_OFFSET_WIZARD_XY_POS uncommented

swissnorp committed:
Update by Request

swissnorp commented:
@qwewer0 
> * I would love to see an LCD message while it probes for the Z reference, so that it isn't just a blank screen.

This is fixed now.
I added a message screen when probing the Z Reference: `Probing Z Reference`.
And an other one for the moving part: `Moving to Probing Pos`

> * And one thing that I don't like, is the homing after the done/cancel button. Is there a particular reason why do we need to home after the wizard?

Yes its absolutely needed. If you don't believe me, try it by uncommenting it 😸. The current_position.z would not be trustable any more under the given sircumstances/if-conditions. But I reduced it to home Z only.

@thinkyhead 
As a cherry on top it would be great if we could extend the messages for homing in marlin so it would be easy to give proper screens:
`MSG_LEVEL_BED_HOMING                = _UxGT("Homing XYZ");`
`MSG_LEVEL_BED_HOMING_XY          = _UxGT("Homing XY");`
`MSG_LEVEL_BED_HOMING_Z          = _UxGT("Homing Z");`

@filippor Thank you for testing and giving a feedback 👍 

I think the code is not perfect but it works. I am relatively new to marlin and I never proberly learned to code 😅. Maybe there is some clean up possible.

Enjoy Testing, norp

swissnorp committed:
Merge branch 'bugfix-2.0.x' into ProbeOffsetWizard-Improvements

qwewer0 commented:
> But I reduced it to home Z only.

That is the same as it was for probe users.

Why can't `do_z_clearance();` be used after the wizard? I changed `queue.inject_P(PSTR("G28Z"));` to `do_z_clearance(20.0);` and it worked just fine. Is there something that I don't know that wouldn't allow us to use `do_z_clearance();`?

qwewer0 commented:
Would it not work?
```
ACTION_ITEM(MSG_BUTTON_DONE, []{
    set_offset_and_go_back(calculated_z_offset);
    do_z_clearance(20.0);
  });

ACTION_ITEM(MSG_BUTTON_CANCEL, []{
    set_offset_and_go_back(z_offset_backup);
    do_z_clearance(20);
  });
```

swissnorp commented:
@qwewer0 Thank you for partizipation. I realy apreciate your skepsis.

> Sorry if I'm missing something.

Yep you are 😉 .
Homing for the wizard was done with `PROBE_OFFSET_START` or with (let's expext) wrong `probe.offset.z`. Changing or resetting the `probe.offset.z` to the proper value does not change `current_position.z`!

**So rehoming Z is absolutely needed!**

> 
> 
> Would it not work?
> 
> ```
> ACTION_ITEM(MSG_BUTTON_DONE, []{
>     set_offset_and_go_back(calculated_z_offset);
>     do_z_clearance(20.0);
>   });
> 
> ACTION_ITEM(MSG_BUTTON_CANCEL, []{
>     set_offset_and_go_back(z_offset_backup);
>     do_z_clearance(20.0);
>   });
> ```

NO 😏  

Did you measure the distance between nozzle and bed, and did it match with the current_position.z?
Make sure that `PROBE_OFFSET_START` or your currently set `probe.offset.z` (from Settings/EEPROM) is more than 1 mm off from 
your proper/real `probe.offset.z`, and measure with a ruler!

Adios, norp

qwewer0 commented:
Tested it with do_z_clearance(20.0);, and after the wizard the LCD showed 20, and the nozzle was at 18.75 with a z offset of -1.25.

Despite all this, I still believe that it should only move the z axis up a little for clearance, and either disengage the steppers or just set the z axis or all to be unhomed/untrusted.

swissnorp commented:
> 
> 
> Tested it with do_z_clearance(20.0);, and after the wizard the LCD showed 20, and the nozzle was at 18.75 with a z offset of -1.25.
> 
> Despite all this, I still believe that it should only move the z axis up a little for clearance, and either disengage the steppers or just set the z axis or all to be unhomed/untrusted.
> Edit: With perfect z offset before the wizard.

But you have had enabled `PROBE_OFFSET_START` with a different value, haven't you? And thats the reason for me to rehome (or to set Z-Axis untrusted, as you mentioned).


**Conclusion:**
- So we have different opinions about homing after the wizard. I vote for rehoming, because I think it's safer and more convenient. And there are several newer and older features, which do a rehoming after a calibration process. I leave it in there!
- But nevertheless I think about adding a `do_z_clearance(Z);` indepentently.
- I will also add a `SanityCheck` to check if `PROBE_OFFSET_START` or `NOZZLE_TO_PROBE_OFFSET.Z`, because the Probe will never trigger at a higher point as the nozzle!

I will work on it tomorrow night (MEZ/CET).

Good night, norp

qwewer0 commented:
If the homing wins, then I'm 100% with it.
And for the a Z clearance after the homing, I think it definitely needs a bit (10-20 mm) of a Z clearance even after the homing, because while the `Z_CLEARANCE_DEPLOY_PROBE` could be big enough, like with a BLTouch setup, but it might be too small with an inductive or laser probe, and we don't want accidental heated bed burning with the wizard, do we?

swissnorp commented:
> 
> 
> If the homing wins, then I'm 100% with it.
> And for the a Z clearance after the homing, I think it definitely needs a bit (10-20 mm) of a Z clearance even after the homing, because while the `Z_CLEARANCE_DEPLOY_PROBE` could be big enough, like with a BLTouch setup, but it might be too small with an inductive or laser probe, and we don't want accidental heated bed burning with the wizard, do we?

I think you are meaning do_z_clearance(Z); before homing, aren't you? After homing it would make absolutely no sens!?

thinkyhead commented:
We have `Z_AFTER_HOMING` to set the desired raise after G28 is complete.

swissnorp committed:
Update menu_probe_offset.cpp

swissnorp commented:
@thinkyhead yes excactly


> **Conclusion:**
> 
>     * So we have different opinions about homing after the wizard. I vote for rehoming, because I think it's safer and more convenient. And there are several newer and older features, which do a rehoming after a calibration process. I leave it in there!
> 
>     * But nevertheless I think about adding a `do_z_clearance(Z);` indepentently.
> 
>     * I will also add a `SanityCheck` to check if `PROBE_OFFSET_START` or `NOZZLE_TO_PROBE_OFFSET.Z`, because the Probe will never trigger at a higher point as the nozzle!

SanityCheck still has to be done.

For the rest, I did it like this:

```
  ACTION_ITEM(MSG_BUTTON_DONE, []{
    set_offset_and_go_back(calculated_z_offset);

    // Set Z to 0, as we can expect
    current_position.z = 0;
    planner.synchronize();

    // Raise Z as if it was homed
    do_z_clearance(0
    #ifdef Z_AFTER_HOMING
      +Z_AFTER_HOMING
    #elseif defined(Z_HOMING_HEIGHT)
      +Z_HOMING_HEIGHT
    #endif
    );
  });

  ACTION_ITEM(MSG_BUTTON_CANCEL, []{
    set_offset_and_go_back(z_offset_backup);

    // If wizard-homing was done with PROBE_OFFSET_WIZARD_START_Z by probe.
    // Set Z to 0 (if higher than 0) to be sure for G28 clearance and rehome Z with backed up offset.
    #if defined(PROBE_OFFSET_WIZARD_START_Z) && EITHER(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN, USE_PROBE_FOR_Z_HOMING)
      if(current_position.z > 0) { current_position.z = 0; planner.synchronize(); }
      queue.inject_P(PSTR("G28Z"));

    // Otherwise do a Z clearance move like after Homing
    #else
      do_z_clearance(0
      #ifdef Z_AFTER_HOMING
        +Z_AFTER_HOMING
      #elseif defined(Z_HOMING_HEIGHT)
        +Z_HOMING_HEIGHT
      #endif
      );
    #endif
  });
```

Good Night, norp

swissnorp committed:
fix last commit

swissnorp committed:
fix the fix :stuck_out_tongue_winking_eye:

qwewer0 commented on Marlin/src/lcd/menu/menu_probe_offset.cpp:
```diff
@@ -107,47 +90,122 @@ void probe_offset_wizard_menu() {
+    sync_plan_position();
+
+    // Raise Z as if it was homed
+    do_z_clearance(0
```
```
// Raise Z
do_z_clearance(20.0	
#ifdef Z_AFTER_HOMING
   - 20.0 + Z_AFTER_HOMING
#elif defined(Z_HOMING_HEIGHT)
   + Z_HOMING_HEIGHT
#endif
);
```

qwewer0 replied:
Of course it applies to the cancel part too.

thinkyhead replied:
`#elif`

qwewer0 replied:
@thinkyhead You are right.

qwewer0 reviewed (changes requested):
It should be safe even if `Z_AFTER_HOMING` and `Z_HOMING_HEIGHT` isn't defined, and move the nozzle up at the end.

qwewer0 commented:
With the latest changes, after done, the nozzle just stays in that position, potentially burning the bed.

qwewer0 commented on Marlin/src/lcd/menu/menu_probe_offset.cpp:
```diff
@@ -107,47 +90,122 @@ void probe_offset_wizard_menu() {
+    // If wizard-homing was done with PROBE_OFFSET_WIZARD_START_Z by probe.
+    // Set Z to 0 (if higher than 0) to be sure for G28 clearance and rehome Z with backed up offset.
+    #if defined(PROBE_OFFSET_WIZARD_START_Z) && EITHER(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN, USE_PROBE_FOR_Z_HOMING)
+      if(current_position.z > 0) { current_position.z = 0; sync_plan_position();; }
```
Extra semicolon?

thinkyhead committed:
Current Z will be trounced

thinkyhead committed:
Update menu_probe_offset.cpp

thinkyhead committed:
Some (but not all) cleanup

qwewer0 commented on Marlin/src/lcd/menu/menu_probe_offset.cpp:
```diff
@@ -107,47 +90,96 @@ void probe_offset_wizard_menu() {
+      do_z_clearance(Z_AFTER_HOMING);
+    #elif defined(Z_HOMING_HEIGHT)
+      do_z_clearance(Z_HOMING_HEIGHT);
+    #endif
```
#ifdef Z_AFTER_HOMING
      do_z_clearance(Z_AFTER_HOMING);
    #elif defined(Z_HOMING_HEIGHT)
      do_z_clearance(Z_HOMING_HEIGHT);
    #else
      do_z_clearance(20.0);
    #endif

qwewer0 reviewed (changes requested):
How about adding an else do_z_clearance?

swissnorp committed:
Nozzle is at z_offset_ref, not 0 (but may be 0)

After Calibrate:
current_position.z = z_offset_ref;

swissnorp commented on Marlin/src/lcd/menu/menu_probe_offset.cpp:
```diff
@@ -99,30 +99,22 @@ void probe_offset_wizard_menu() {
     // If wizard-homing was done with PROBE_OFFSET_WIZARD_START_Z by probe.
-    // Set Z to 0 (if higher than 0) to be sure for G28 clearance and rehome Z with backed up offset.
-    #if defined(PROBE_OFFSET_WIZARD_START_Z) && EITHER(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN, USE_PROBE_FOR_Z_HOMING)
-      if(current_position.z > 0) { current_position.z = 0; sync_plan_position();; }
```
I realy think we need this. Because if the Nozzle is close to the bed but the current_position.z is too high, there may will not be enough clearance for probing! We may can not trust `PROBE_OFFSET_WIZARD_START_Z`.
`if(current_position.z > 0) { current_position.z = 0; sync_plan_position(); }`

thinkyhead replied:
I believe the better solution is to use `set_axis_never_homed(Z_AXIS)` and there is already precedent for that.

swissnorp committed:
add do_z_clearance anyway

Because there are many default configs with neither defined `Z_AFTER_HOMING` nor `Z_HOMING_HEIGHT`.
They may home to the Z max endstop.

swissnorp committed:
readd CANCEL/Probe/current.z = 0

I realy think we need this. Because if the Nozzle is close to the bed but the current_position.z is too high, there may will not be enough clearance for probing! We may can not trust PROBE_OFFSET_WIZARD_START_Z.

swissnorp committed:
Update SanityCheck.h

There should be no probe that triggers after the nozzle has hit the bed. So these values should be negative!

swissnorp committed:
Fix SanityCheck.h

NOZZLE_TO_PROBE_OFFSET.Z must be allowed :smile:

swissnorp committed:
Fix the fix?

swissnorp committed:
Merge branch 'bugfix-2.0.x' into ProbeOffsetWizard-Improvements

swissnorp committed:
Update SanityCheck.h

Allow minor positive Value

swissnorp commented:
SanityCheck may needs some tinkering... Ther could be conflicts with probing mechanisms that use the nozzle as a probe beeing pushed up and activating a switch!

For the rest: I think that's it. Besides that it needs some testing, reviewing and cleanup.

Good night, norp

thinkyhead committed:
Tweak sanity, raise, Z homing

thinkyhead committed:
More sanity tweaks

thinkyhead committed:
noun

ohrn commented:
Hi!

This sanity check addition breaks compilation on my printer, a ZYYX Pro.

```
    static_assert(sanity_nozzle_to_probe_offset.z <= 0.25,
                  "Are you sure your Probe triggers above the nozzle? Set a negative Z value in the NOZZLE_TO_PROBE_OFFSET.");
```

The probe truly triggers 2 mm "above" the nozzle. The printer has a micro switch next to the nozzle and notches cut out of the build plate edges. When probing the nozzle lowers into a notch and the switch engages on the plate.

Z safe probing and carefully defined probe points matching the notches is  a must on this printer! :smile: 

I guess this is not a common design, but it would be nice if this error could be turned into a warning, or another "I really know what I'm doing" config added to allow a positive z offset.

/Fredrik

swissnorp commented:
@ohrn Thats some crazy endstop mount 😄 .
Could you test https://github.com/MarlinFirmware/Marlin/pull/20344/commits/97021703d104024f1939829d7cfc35dc432b6dd0 and give some feedback? Make sure you enable/uncomment `NOZZLE_TO_PROBE_OFFSET_Z_POSITIVE`.

Thanks for your input.
